client/src/components/DisplayTable/DisplayTable.js

```diff
@@ -113,6 +122,37 @@ export class DisplayTable extends React.Component {
     }
   }
 
+  change_page_size = (new_page_size) =>
+    this.setState(
+      {
+        loading: true,
+      },
+      _.debounce(() =>
+        this.setState({
+          page_size: new_page_size,
+          current_page: 0,
+          loading: false,
+        })
+      )
+    );
+
+  change_page = (page) =>
+    this.state.current_page !== page &&
+    this.setState(
+      {
+        loading: true,
+      },
+      _.debounce(() =>
+        this.setState(
+          {
+            current_page: page,
+            loading: false,
+          },
+          _.debounce(() => this.first_data_ref.current.focus()) //for some reason if i don't debounce it by 0ms, it will sometimes execute without a proper focus
```

Comment:
Yeah, React will call your `setState` callback before rendering but I believe React calls these callbacks asynchronously (so that it doesn't have to block rendering). Microtiming dependent, I think the callback could be executed either before OR after the resulting render is dome. When it happens before the render finishes, your focus probably went on to an element that was immediately removed from the DOM by the render.

Putting the callback itself in a debounce (indirect setTimeout), even with 0ms, means it's shunted back in to the task queue a second time. If the timing was close enough that it it was sometimes called before and sometimes after the render finished to start with, then it'd make sense that bumping it back in to the queue a second time would be good enough to let the render finish first (I don't know, maybe if you tried enough times you'd still not have the desired behaviour, it's out of your hands when messing with async like this). 

Read up on the JS event loop for a better understanding of how your code's being executed. If you don't find any explanations that click with you I can try and explain it my self some time. 

Anyway, as I said above, this isn't a good use case for `setState` callbacks (my bias perhaps, but I think there's almost no good use cases for them). I've once or twice had to use 0ms timeouts to try and trick the JS engine in to executing how I wanted, but it's kind of an absolute last-ditch hack. In a case like this where it's possible to make the order of events linear with a little bit of extra state, always go for that.

Reply:
Yea I initially also didn't want to use the `setState` callbacks either but for some reason it was the only way for me to be able to almost consistently have code run after the dom updates instead of the virtual dom.

I mean, OTOH it might just be my mac that has this issue since it didn't seem like anyone else had it. I can give `componentDidUpdate` a try again and see if it works for others

Reply:
One of the official use cases for `componentDidUpdate` is inspecting and operating on the DOM, it's not supposed to be called until after the actual DOM has been updated 🤔